zebra/zebra_router.h

```diff
@@ -36,6 +37,36 @@ extern "C" {
  * perspective.
  */
 
+#ifdef HAVE_SCRIPTING
+/*
+ * Script name hash
+ */
+PREDECL_HASH(script_names);
+
+struct script_names_entry {
+	/* Name of a Lua hook call */
+	char function_name[MAXPATHLEN];
+
+	/* Lua script in which to look for it */
+	char script_name[MAXPATHLEN];
+
+	struct script_names_item item;
+};
+
+void zebra_script_init(struct script_names_head *script_names_hash);
+
+int zebra_script_hash_cmp(const struct script_names_entry *snhe1,
+			  const struct script_names_entry *snhe2);
+uint32_t zebra_script_hash_key(const struct script_names_entry *snhe);
+
+int zebra_script_insert_script_name(struct script_names_head *script_names_hash,
+				    char *function_name, char *script_name);
+
+DECLARE_HASH(script_names, struct script_names_entry, item,
+	     zebra_script_hash_cmp, zebra_script_hash_key);
+
+#endif /* HAVE_SCRIPTING */
```

Comment:
So... presumably other daemons are gonna need some facility like this other than just zebra. Is it possible for us to abstract most of this into lib/ so a daemon can more easily add this functionality itself?

Look at `lib/nexthop_group.c`  and it's `vty` commands are an example of this. Basically a daemon can register callbacks to handle cli commands from the user. This way you create the cli in lib, and each daemon can maintain it's own hash table of the function names it's interested in.

Reply:
Marking resolved since we went the route of keeping individual hookpoint calls but moving all the hashing/table into lib/.